File: frontend/src/views/stationview/monitoring/stationFeedsView/feedUse.ts
/*
 *     SPDX-License-Identifier: AGPL-3.0-only
 *
 *     Copyright (C) RainbowDashLabs and Contributor
 */
import type {FeedUse} from '@/api/feedToken'

const ACTIVE_WINDOW_MS = 7 * 24 * 3600_000

/** The later of the two fetch stamps, or null where neither feed has ever been fetched. */
export function lastFetched(use: FeedUse): string | null {
    const stamps = [use.icalPolledAt, use.notificationPolledAt].filter((s): s is string => !!s)
    if (stamps.length === 0) return null
    return stamps.reduce((latest, stamp) => (stamp > latest ? stamp : latest))
}

/** Whether anything has fetched either feed within the last week. */
export function usedRecently(use: FeedUse, now = Date.now()): boolean {
    const stamp = lastFetched(use)
    return !!stamp && now - new Date(stamp).getTime() <= ACTIVE_WINDOW_MS
}

File: src/main/java/dev/chojo/ember/feature/feed/entity/FeedUse.java
/*
 *     SPDX-License-Identifier: AGPL-3.0-only
 *
 *     Copyright (C) RainbowDashLabs and Contributor
 */
package dev.chojo.ember.feature.feed.entity;

import de.chojo.sadu.mapper.rowmapper.RowMapping;

import java.time.Instant;

import static de.chojo.sadu.queries.converter.StandardValueConverter.INSTANT_TIMESTAMP;

/**
 * One member's standing subscription, as the station sees it.
 *
 * <p>Deliberately without the token. It is the whole key to that person's calendar, and a page that
 * lists it is a page from which somebody else's appointments can be read. What a station needs to
 * know is who has a subscription and whether anything ever fetches it, and neither question needs
 * the secret.
 *
 * @param memberId             the member the subscription belongs to
 * @param createdAt            when they set it up
 * @param icalPolledAt         when a calendar last fetched it, or null where none ever has
 * @param notificationPolledAt when a reader last fetched the notifications, or null where none has
 */
public record FeedUse(int memberId, Instant createdAt, Instant icalPolledAt, Instant notificationPolledAt) {
    public static RowMapping<FeedUse> map() {
        return row -> new FeedUse(
                row.getInt("member_id"),
                row.get("created_at", INSTANT_TIMESTAMP),
                row.get("ical_polled_at", INSTANT_TIMESTAMP),
                row.get("notification_polled_at", INSTANT_TIMESTAMP));
    }
}
